### PR TITLE
[Raft] Add some ut, fix some bugs.

### PR DESCRIPTION
**Pull request by critical27**
Add some raft ut, add some methods to make test easy. There are some bugs in my submission:
1. When followers receives appendLogRequest, we will check whether leader's term is bigger than current term in verifyLeader. Suppose the scenario: leader and followers commit some logs, and one follower crash then reboot, its current term is last wal term, and it won't accept leader's log any more. So I think leader's term greater than or equal to current term is ok.
2. When followers receives appendLogRequest, we will try to commit logs from committedLogId_ + 1 to leader's commit log id, but it seems we can't guarantee that we have that log. So we can only commit to min(lastLogId_, leader's commit log ig).

### CODE REVIEW

**Comment by dangleptr** [comment]
Well done.  It is a good starting for raft tests.

**Comment by critical27** [comment]
Jenkins go

File: src/kvstore/raftex/test/LogAppendTest.cpp [hunk 1]
```diff
@@ -29,32 +29,17 @@ TEST(LogAppend, SimpleAppendWithOneCopy) {
     std::vector<HostAddr> allHosts;
     std::vector<std::shared_ptr<RaftexService>> services;
     std::vector<std::shared_ptr<test::TestShard>> copies;
+    std::vector<LogID> lastCommittedLogId;
```
Comment (dangleptr) [inline src/kvstore/raftex/test/LogAppendTest.cpp]: What's the purpose?

File: src/kvstore/raftex/RaftPart.cpp [hunk 2]
```diff
@@ -594,7 +596,8 @@ void RaftPart::replicateLogs(folly::EventBase* eb,
                    prevLogTerm,
                    pHosts = std::move(hosts)] (folly::Try<AppendLogResponses>&& result) mutable {
             VLOG(2) << self->idStr_ << "Received enough response";
-            CHECK(!result.hasException());
+            // TODO: since we kill some peers in ut, this may fail, handle exception if necessary
+            // CHECK(!result.hasException());
```
[the remark this answers is not in the document]
Comment (critical27) [reply src/kvstore/raftex/RaftPart.cpp]: I mock disconnect from one Host, and remove it from its peer, and it may throw exception like `runtime_error: Not enough futures` because of the Host has been destroyed.

File: src/kvstore/raftex/RaftPart.cpp [hunk 3]
```diff
@@ -1228,7 +1235,7 @@ cpp2::ErrorCode RaftPart::verifyLeader(
     }
 
     // Make sure the remote term is greater than local's
-    if (req.get_current_term() <= term_) {
+    if (req.get_current_term() < term_) {
```
[the remark this answers is not in the document]
Comment (critical27) [reply src/kvstore/raftex/RaftPart.cpp]: Yeah, it has been handled before when we check the leadership stickness

File: src/kvstore/raftex/test/LogAppendTest.cpp (hunk 1 above)
Comment (critical27) [reply src/kvstore/raftex/test/LogAppendTest.cpp]: It is used to store a copy's last commit log id outside the Part itself (just like the kLastCommittedIdKey we store it in rocksdb). When a TestShard reboot we can get lastCommittedLogId here.

**Comment by critical27** [comment]
> We need fault injector to mock some errors. Let's design it carefully offline.

Sure, I have not came up with how to mock those errors.

File: src/kvstore/raftex/RaftPart.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/kvstore/raftex/RaftPart.cpp]: if you want to mock host disconnect,  you should stop the host directly instead of removing it from others's peers.

File: src/kvstore/raftex/RaftPart.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/kvstore/raftex/RaftPart.cpp]: Not exactly,  in your code,  if a new leader with the same term comes,   the follower will still follow it.

File: src/kvstore/raftex/test/LogAppendTest.cpp (hunk 1 above)
Comment (dangleptr) [reply src/kvstore/raftex/test/LogAppendTest.cpp]: If so, why not put it in TestShard?

File: src/kvstore/raftex/RaftPart.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (critical27) [reply src/kvstore/raftex/RaftPart.cpp]: Em, not only term, but it also check req.get_leader_ip() == leader_.first and req.get_leader_port() == leader_.second. I don't quite get your point. If a new leader comes with the same term, if it is the same leader, everything is fine. But I don't think we can have two leader with same term.

File: src/kvstore/raftex/test/LogAppendTest.cpp (hunk 1 above)
Comment (critical27) [reply src/kvstore/raftex/test/LogAppendTest.cpp]: Good suggestion

File: src/kvstore/raftex/RaftPart.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (critical27) [reply src/kvstore/raftex/RaftPart.cpp]: Yeah, you're right. It only throw the exception when I modify the peers. But do you think it is necessary to mock network partition?

[the remark this answers is not in the document]
Comment (dangleptr) [reply src/kvstore/raftex/RaftPart.cpp]: Of course, we will mock the network partition situation.  We could leave the job in our raft test framework.

File: src/kvstore/raftex/RaftPart.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dangleptr) [reply src/kvstore/raftex/RaftPart.cpp]: We can't assume it never happen,   maybe some bugs inside it.

File: src/kvstore/raftex/test/RaftCase.cpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
+/* Copyright (c) 2018 vesoft inc. All rights reserved.
```
Comment (dangleptr) [inline src/kvstore/raftex/test/RaftCase.cpp]: 2019

File: src/kvstore/wal/FileBasedWal.cpp [hunk 5]
```diff
@@ -602,7 +602,6 @@ bool FileBasedWal::rollbackToLog(LogID id) {
         }
 
         auto it = walFiles_.upper_bound(id);
```
Comment (dangleptr) [inline src/kvstore/wal/FileBasedWal.cpp]: Separate the changes into another pr.

File: src/kvstore/raftex/RaftPart.cpp [hunk 6]
```diff
@@ -716,17 +718,6 @@ void RaftPart::processAppendLogResponses(
         if (!iter.empty()) {
             appendLogsInternal(std::move(iter), currTerm);
         }
-    } else {
```
Comment (dangleptr) [inline src/kvstore/raftex/RaftPart.cpp]: Use another pr for it.

File: src/kvstore/wal/FileBasedWal.cpp (hunk 5 above)
Comment (critical27) [reply src/kvstore/wal/FileBasedWal.cpp]: ok

File: src/kvstore/raftex/RaftPart.cpp (hunk 6 above)
Comment (critical27) [reply src/kvstore/raftex/RaftPart.cpp]: ok
